[DOC] Adding MWE to some docstrings

thatlittleboy opened the pull request:
# PR Description

Some effort towards adding and/or standardizing the MWE in function docstrings (#968 , #972). I've tried not to touch other things unrelated to this specific goal, to keep this PR more focused.
With one exception, which is the `groupby_agg` function -> I've just kept 2 minimal examples within the docstring and moved the rest of the detailed explanation to the example notebook.

Please describe the changes proposed in the pull request:
- Added MWE to and cleaned up `functions/rename_columns.py`
- Added MWE to and cleaned up `functions/reorder_columns.py`
- Added MWE to and cleaned up `functions/to_datetime.py`
- Added MWE to and cleaned up `functions/also.py`
- Added MWE to and cleaned up `functions/then.py`
- Added MWE to and cleaned up `functions/dropnotnull.py`
- Added MWE to and cleaned up `functions/groupby_agg.py` & associated example notebook.

**This PR is related to #972 .**


(If we have some best practices regarding committing ipynb files into the repo, please educate me on it as well, thanks!)

# PR Checklist

<!-- This checklist exists for newcomers who are not yet familiar with our requirements. If you are experienced with
the project, please feel free to delete this section. -->

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.
<!-- Doing this helps us keep the commit history much cleaner than it would otherwise be. -->
2. [x] If you're not on the contributors list, add yourself to `AUTHORS.rst`.
<!-- We'd like to acknowledge your contributions! -->
3. [ ] Add a line to `CHANGELOG.md` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

thatlittleboy committed:
Update rename_columns docstring with MWE

thatlittleboy committed:
Update reorder_columns docstring with MWE

thatlittleboy committed:
Update to_datetime docstring with MWE

thatlittleboy committed:
Update then docstring with MWE

thatlittleboy committed:
Update also docstring with MWE

thatlittleboy committed:
Update dropnotnull docstring with MWE

thatlittleboy committed:
Update groupby_agg docstring with MWE

Clean up docstring examples, shifting some of the heavier explanation to the examples notebook.
Also update the type hint of `by` slightly (though still partially incomplete).

thatlittleboy committed:
Fix lint

thatlittleboy committed:
Update flag_nulls docstring with MWE

thatlittleboy committed:
Fix lint

ericmjl commented:
This is really good work, @thatlittleboy! Thank you for handling this first pass.

I noticed that primarily it is the code style checks that are failing.

Here's the output to help:

https://github.com/pyjanitor-devs/pyjanitor/runs/4692357467?check_suite_focus=true#step:4:58

As for codecov/project, something's off. Can I confirm that you have merged latest `dev` into `enh/mwe-sprint`?

Finally, w.r.t. notebooks, basically the CI checks that there are no outputs in the notebooks. I haven't had the bandwidth to review notebooks in-depth alone, so I'd love help here - primarily to ensure that they read well. That said, I think notebooks will go into the `pyjanitor-examples` repo instead, so we don't have to worry about them here.

thatlittleboy commented:
@ericmjl Yeah I seem to be having trouble with ignoring both DAR402 (raising extraneous exceptions) from darglint and E501 (line too long) from flake8 at the same time. E.g. doing `# noqa: DAR402,E501` just outside the docstring causes the darglint run `darglint -v 2 ...` to fail.. But actually running `flake8 ...` is fine. See second image below.

Actually, do we need to run darglint checks explicitly in the CI? It seems that flake8 already incorporates the darglint checks just fine (see: running flake8 throws DAR402 linting error

<img width="767" alt="image" src="https://user-images.githubusercontent.com/30731072/147949815-9d21a9d3-71e4-48e9-9152-60f3f8e214fa.png">

What's more, flake8 actually respects the noqa specifications, but not darglint..

<img width="767" alt="image" src="https://user-images.githubusercontent.com/30731072/147949992-e835cc48-5ef5-4acc-aae2-2387df990a82.png">

Anyway, it's not a big deal, I'll just adjust the docstrings so E501 won't complain.

----------

Regarding latest changes, yeah, I think so, here's the graph 

<img width="661" alt="image" src="https://user-images.githubusercontent.com/30731072/147950331-a6ee205c-3448-443c-98d8-f03a3bfae795.png">

`enh/mwe_sprint` is branched off from the latest merged PR in upstream.

---------

Okay, thanks for the heads up. I'll clear the output of the notebook then. I'll still leave the notebook edits in this PR, we could just copy it over later.

thatlittleboy committed:
Fix lint

Struggling with darglint and flake8 running at the same time
because darglint and flake8 don't seem to respect the noqa at the
same time

thatlittleboy committed:
Remove output from notebook

thatlittleboy commented:
Seems like it is just codecov failing, not quite sure why it's trying to compare to d6ef2c3. I'll remove the draft status for now,

samukweku commented on janitor/functions/dropnotnull.py:
```diff
@@ -8,16 +8,22 @@
+        >>> import numpy as np
+        >>> import pandas as pd
+        >>> import janitor
+        >>> df = pd.DataFrame({"a": [1., np.NaN, 3.], "b": [None, "y", "z"]})
```
Not a bad idea to show the original dataframe as well. What do u think?

ericmjl replied:
Yes, it's a good idea.

samukweku commented on janitor/functions/rename_columns.py:
```diff
@@ -8,31 +8,26 @@
-    This is just syntactic sugar/a convenience function
-    for renaming one column at a time.
+    This is just syntactic sugar/a convenience function for renaming one column at a time.
     If you are convinced that there are multiple columns in need of changing,
```
Should we remove this line, since rename columns fits in for multiple columns. There is also the larger discussion about dropping the singular forms, since rename columns covers single use as well . Pinging @ericmjl

ericmjl replied:
Yes, good point, @samukweku. Let's delete this line.

ericmjl commented:
@thatlittleboy your comment:

> not quite sure why it's trying to compare to d6ef2c3

gave me enough information to try another debug. I've pushed one small change directly to `dev`, in which we re-run all tests and upload a definitive `dev` branch test coverage file to Codecov.io. I believe that this should fix our future coverage issues.

ericmjl commented:
@thatlittleboy I was checking for the docs site preview and I realize now that if you use a fork, you won't have access to my netlify API keys which are necessary for deploying the website preview. For future PRs, would you like to branch off the main repo instead of using a personal fork? That will enable you to see previews.

Finally, once I see the CI system finish up on `dev`, I'd encourage you to do some commits (whether empty or with changes), to re-trigger the CI build here.

thatlittleboy commented:
Got it, and well-received on all fronts. I'll make the necessary changes later tonight. Thanks both.

thatlittleboy committed:
Merge branch 'pyjanitor-devs:dev' into enh/mwe-sprint

thatlittleboy committed:
Show outputs of initial df, particularly for inplace changes

thatlittleboy committed:
Remove note about rename_column

thatlittleboy commented:
Ok I've made the necessary changes. Just 2 final comments from me:
1. Does this warrant a CHANGELOG update? If required, I can add it in as well.
2. Sorry about branching off a fork and having no docs preview here. Will keep this in mind for future PRs. If it's more convenient for any further review by either of you, I'll be happy to close this one and open a new one in its place to merge in from `pyjanitor-devs:enh/mwe-sprint` instead.

samukweku commented:
@thatlittleboy it seems the cells in the notebook were not executed and as such the dataframe outputs are not displayed

thatlittleboy commented:
@samukweku Yeah, I deleted them. Eric mentioned

> Finally, w.r.t. notebooks, basically the CI checks that there are no outputs in the notebooks. I haven't had the bandwidth to review notebooks in-depth alone, so I'd love help here - primarily to ensure that they read well. That said, I think notebooks will go into the pyjanitor-examples repo instead, so we don't have to worry about them here.

So I thought I had to remove the outputs to prevent the CI from failing.

ericmjl commented:
> Does this warrant a CHANGELOG update? If required, I can add it in as well.

Let's do this in the next PR, a generic line should do the trick.

---

We're good to go, I'm merging! :tada:
